.github/workflows/_build-and-test-locally.yml

```diff
@@ -334,24 +344,127 @@ jobs:
         with:
           submodules: true
 
+      # Needed for detecting new and updated tests
+      - uses: actions/checkout@v4
+        if: github.event.pull_request.head.sha && inputs.test-selection == '' && matrix.pg_version == 'v17'
+        with:
+          submodules: false
+          clean: false
+          fetch-depth: 1000
+
+      - name: Cache poetry deps
+        if: github.event.pull_request.head.sha && inputs.test-selection == '' && matrix.pg_version == 'v17'
+        uses: actions/cache@v4
+        with:
+            path: ~/.cache/pypoetry/virtualenvs
+            key: v2-${{ runner.os }}-${{ runner.arch }}-python-deps-bookworm-${{ hashFiles('poetry.lock') }}
+
+      - name: Install Python deps
+        if: github.event.pull_request.head.sha && inputs.test-selection == '' && matrix.pg_version == 'v17'
+        shell: bash -euxo pipefail {0}
+        run: ./scripts/pysync
+
+      - name: Detect new and updated tests
+        id: select_tests_to_test
+        if: github.event.pull_request.head.sha && inputs.test-selection == '' && matrix.pg_version == 'v17'
+        env:
+          COMMIT_SHA: ${{ github.event.pull_request.head.sha || github.sha }}
+          BASE_SHA: ${{ github.event.pull_request.base.sha || github.sha }}
+        run: |
+          MAX_TESTS=3
```

Comment:
This `MAX_TESTS=3` makes me a bit unease.
Why do we test only 3 random changed tests?
I think we need to test all of them and add a label to allow skipping this behaviour (if someone decide to update style in all tests, for example)